service/importer/service_file.go

```diff
@@ -4,70 +4,121 @@ import (
-	"github.com/xeipuuv/gojsonschema"
+	"github.com/go-playground/locales/en"
+	ut "github.com/go-playground/universal-translator"
+	"github.com/mesg-foundation/core/x/xvalidator"
+	validator "gopkg.in/go-playground/validator.v9"
+	en_translations "gopkg.in/go-playground/validator.v9/translations/en"
 	yaml "gopkg.in/yaml.v2"
 )
 
-// ConfigurationDependencyKey is the reserved key of the service's configuration in the dependencies array.
-const ConfigurationDependencyKey = "service"
+const (
+	namespacePrefix = "ServiceDefinition."
+)
 
 func readServiceFile(path string) ([]byte, error) {
 	file := filepath.Join(path, "mesg.yml")
 	return ioutil.ReadFile(file)
 }
 
 // validateServiceFile returns a list of warnings.
-func validateServiceFile(data []byte) ([]string, error) {
-	var body interface{}
-	if err := yaml.Unmarshal(data, &body); err != nil {
-		return nil, fmt.Errorf("parse mesg.yml error: %s", err)
-	}
-	body = convert(body)
-	result, err := validateServiceFileSchema(body, "service/importer/assets/schema.json")
-	if err != nil {
-		return nil, err
-	}
-	var warnings []string
-	if !result.Valid() {
-		for _, warning := range result.Errors() {
-			warnings = append(warnings, warning.String())
+func validateServiceFile(data []byte) []string {
+	var service ServiceDefinition
+	if err := yaml.UnmarshalStrict(data, &service); err != nil {
+		errs, ok := err.(*yaml.TypeError)
+		if !ok {
+			return []string{err.Error()}
 		}
+		return errs.Errors
 	}
-	if depKeyWarning := validateServiceFileDependencyKey(body); depKeyWarning != "" {
-		warnings = append(warnings, depKeyWarning)
-	}
-	return warnings, nil
+	return validateServiceStruct(&service)
 }
 
-func validateServiceFileDependencyKey(data interface{}) (warning string) {
-	s, _ := data.(map[string]interface{})
-	dep, _ := s["dependencies"].(map[string]interface{})
-	if dep[ConfigurationDependencyKey] != nil {
-		return fmt.Sprintf("cannot use %q as dependency key", ConfigurationDependencyKey)
+func validateServiceStruct(service *ServiceDefinition) []string {
+	validate, trans := newValidator()
+	err := validate.Struct(service)
+	warnings := []string{}
+	if err != nil {
+		errs := err.(validator.ValidationErrors)
+		for _, e := range errs {
+			// Remove the name of the struct and the field from namespace
+			trimmedNamespace := strings.TrimPrefix(e.Namespace(), namespacePrefix)
+			trimmedNamespace = strings.TrimSuffix(trimmedNamespace, e.Field())
+			// Only use it when in-cascade field
+			namespace := ""
+			if e.Field() != trimmedNamespace {
+				namespace = trimmedNamespace
+			}
+			warnings = append(warnings, fmt.Sprintf("%s%s", namespace, e.Translate(trans)))
+		}
 	}
-	return ""
-}
 
-func validateServiceFileSchema(data interface{}, schemaPath string) (*gojsonschema.Result, error) {
-	schemaData, err := assets.Asset(schemaPath)
-	if err != nil {
-		return nil, err
+	for key, dep := range service.Dependencies {
+		if dep == nil {
+			continue
+		}
+		if dep.Image == "" {
```

Comment:
why don't we check this with govalidator's 'required' tag?

Reply:
because it depends on the context, if in a dependency then the image is required, if in the configuration no so it's more complicated to have this with govalidator than just code it